### PR TITLE
[BEAM-1350] BigQuery: refactor services so that all queryConfig happens in BigQueryIO

### PR DESCRIPTION
**Pull request by dhalperi**
By putting all the configuration in the same place, we can avoid
bugs that happen from mismatching code across files.

Also made a few unnecessarily-public APIs package-private.

And improved tests, removed a few dataflow references.

Forward port from Dataflow.

### CODE REVIEW

**Comment by dhalperi** [comment]
R: @peihe

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -1162,8 +1161,9 @@ private void executeQuery(
 
     private JobConfigurationQuery createBasicQueryConfig() {
       return new JobConfigurationQuery()
-          .setQuery(query.get())
           .setFlattenResults(flattenResults)
+          .setPriority("BATCH")
```
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This downgrades priority in dryRun and reader to BATCH.
what is the reason?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java [hunk 2]
```diff
@@ -88,64 +88,54 @@
...
 
   /**
    * Constructs a {@code BigQueryTableRowIterator} that reads from the specified table.
    */
-  public static BigQueryTableRowIterator fromTable(TableReference ref, Bigquery client) {
+  static BigQueryTableRowIterator fromTable(TableReference ref, Bigquery client) {
     checkNotNull(ref, "ref");
     checkNotNull(client, "client");
-    return new BigQueryTableRowIterator(ref, null, ref.getProjectId(), client, true, true);
+    return new BigQueryTableRowIterator(ref, null, ref.getProjectId(), client);
```
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java]: nit: null /* queryConfig */

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java [hunk 3]
```diff
@@ -88,64 +88,54 @@
...
-    checkNotNull(query, "query");
+  static BigQueryTableRowIterator fromQuery(
+      JobConfigurationQuery queryConfig, String projectId, Bigquery client) {
+    checkNotNull(queryConfig, "queryConfig");
     checkNotNull(projectId, "projectId");
     checkNotNull(client, "client");
-    return new BigQueryTableRowIterator(null, query, projectId, client,
-        MoreObjects.firstNonNull(flattenResults, Boolean.TRUE),
-        MoreObjects.firstNonNull(useLegacySql, Boolean.TRUE));
+    return new BigQueryTableRowIterator(null, queryConfig, projectId, client);
```
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java]: nit: null /* ref */

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 1 above)
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Can you say more? This line is moved from L1152 in "startQueryJob".

We should use the same priority everywhere. priority is ignored for dry run.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServices.java [hunk 4]
```diff
@@ -57,8 +57,7 @@
    * Returns a real, mock, or fake {@link BigQueryJsonReader} to query tables.
    */
   BigQueryJsonReader getReaderFromQuery(
-      BigQueryOptions bqOptions, String query, String projectId, @Nullable Boolean flatten,
-      @Nullable Boolean useLegacySql);
+      BigQueryOptions bqOptions, JobConfigurationQuery queryConfig, String projectId);
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServices.java]: Done, but as a general comment I do not need to cleanup pre-existing style issues in this PR.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java [hunk 5]
```diff
@@ -800,15 +799,10 @@ private BigQueryJsonReaderImpl(BigQueryTableRowIterator iterator) {
     }
 
     private static BigQueryJsonReader fromQuery(
-        BigQueryOptions bqOptions,
-        String query,
-        String projectId,
-        @Nullable Boolean flattenResults,
-        @Nullable Boolean useLegacySql) {
+        BigQueryOptions bqOptions, JobConfigurationQuery queryConfig, String projectId) {
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java]: This is a private function with arguments not annotated `@Nullable`, does not need such checks.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java (hunk 2 above)
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java]: done, though this is again fixing a pre-existing issue that need not be addressed here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java (hunk 3 above)
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryTableRowIterator.java]: done, though this is again fixing a pre-existing issue that need not be addressed here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java [hunk 6]
```diff
@@ -100,9 +100,8 @@ public BigQueryJsonReader getReaderFromTable(BigQueryOptions bqOptions, TableRef
 
   @Override
   public BigQueryJsonReader getReaderFromQuery(
-      BigQueryOptions bqOptions, String query, String projectId, @Nullable Boolean flatten,
-      @Nullable Boolean useLegacySql) {
-    return BigQueryJsonReaderImpl.fromQuery(bqOptions, query, projectId, flatten, useLegacySql);
+      BigQueryOptions bqOptions, JobConfigurationQuery queryConfig, String projectId) {
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java]: done, though this is again fixing a pre-existing issue that need not be addressed here.

**Comment by dhalperi** [comment]
The build is broken by Python merge, but PTAL anyway.

**Comment by dhalperi** [comment]
retest this please

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 1 above)
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: createBasicQueryConfig() is also called in createReader(), dryRunQueryIfNeeded(), and executeQuery().
BigQuery defaults to iteractive if priority is not set. So, this PR will change createReader() and dryRunQueryIfNeeded() to batch. And, BATCH queries take minutes to hours to start based on the big query web doc:

batch query starts as soon as idle resources are available, usually within a few minutes. If BigQuery hasn't started the query within 24 hours, BigQuery changes the job priority to interactive.
https://cloud.google.com/bigquery/querying-data

IMO, createReader(), dryRunQueryIfNeeded() benefits from low latency, but executeQuery() doesn't. So, I like the current priority settings, as long as batch queries don't take hours to start.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java (hunk 5 above)
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java]: then the caller in public function should checkNotNull in line 104?

Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java]: I've fixed enough pre-existing issues in this code -- please hold anymore nits in this PR that are not related to new code.

You can file a separate PR to address your nits.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 1 above)
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Reverted this change. I think this will lead to inconsistent behavior across runners which violates user expectation, but I'm tired.

**Comment by peihe** [comment]
LGTM. Thanks Dan!

I will merge after travis pass

**Comment by dhalperi** [comment]
Thanks! merged.
